{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f3fc16ee-4727-4644-8cad-0842385348dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scirpy as ir\n",
    "import muon as mu\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92d5a10d-685b-49ce-8256-e657716e32e3",
   "metadata": {},
   "source": [
    "# Query epitope databases\n",
    "- search for TCRs with specificity annotatation from previous studies.\n",
    "\n",
    "**Databases**\n",
    "- [VDJdb](https://vdjdb.cdr3.net/), a curated database of T-cell receptor (TCR) sequences with known antigen specificities.\n",
    "- [IEDB](https://www.iedb.org/): immune epitope data related to all species studied and includes antibody, T cell, and MHC binding contexts associated with infectious, allergic, autoimmune, and transplant related diseases. IEDB [Fleri et al., 2017]\n",
    "- [McPAS-TCR](http://friedmanlab.weizmann.ac.il/McPAS-TCR/)\n",
    "- [PIRD](https://db.cngb.org/pird/)\n",
    "- [immuneCodeTM](https://clients.adaptivebiotech.com/pub/covid-2020) for covid specific TCRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9efffdd2-1a50-444c-899d-d1f924628c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "    mdata = mu.read_h5mu(\"../data/tcr_data/121c_clone_id_full.h5mu\")\n",
    "\n",
    "    ### Split in B- and T- cells\n",
    "    b_cells = {'B intermediate', 'Plasmablast'}\n",
    "    t_cells = {'gdT', 'CD4 Memory', 'CD8 Memory', 'CD4 Naive', 'CD8 Proliferating', 'CD8 Naive', 'CD4 Proliferating', 'Treg'}\n",
    "    \n",
    "    bcell_mdata = mdata[mdata[\"gex\"].obs[\"consensus_celltype\"].isin(b_cells)].copy()\n",
    "    tcell_mdata = mdata[mdata[\"gex\"].obs[\"consensus_celltype\"].isin(t_cells)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "565bdaeb-972b-4f35-a4b3-c39d792cd689",
   "metadata": {},
   "outputs": [],
   "source": [
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "    iedb = ir.datasets.iedb()\n",
    "    vdjdb = ir.datasets.vdjdb()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c5d0cdd-35bd-4cd3-a9c5-5cf66b29b445",
   "metadata": {},
   "source": [
    "### Look at the HLA molecules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3cb1a72e-9117-43de-ba20-17b90edaee9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mhc.a\n",
       "HLA-A*02       14880\n",
       "HLA-A*03:01    14714\n",
       "HLA-A*02:01    10521\n",
       "HLA-A*11:01     2942\n",
       "Mamu-A*01       2119\n",
       "HLA-B*07:02     1981\n",
       "HLA-B*08:01     1693\n",
       "H-2Kb           1691\n",
       "HLA-A*01:01     1583\n",
       "HLA-DRA*01      1414\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vdjdb.obs.value_counts(\"mhc.a\").iloc[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4d91651-c516-4d7a-9f62-f15a4bedb508",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "antigen.species\n",
      "CMV            23590\n",
      "InfluenzaA     11194\n",
      "EBV             6661\n",
      "SARS-CoV-2      6237\n",
      "HomoSapiens     4131\n",
      "HIV-1           3408\n",
      "SIV             2119\n",
      "HCV             1559\n",
      "MCMV             923\n",
      "YFV              542\n",
      "Name: count, dtype: int64\n",
      "**************************************************\n",
      "There are 42 antigen species\n"
     ]
    }
   ],
   "source": [
    "print(vdjdb.obs.value_counts(\"antigen.species\").iloc[:10])\n",
    "print(\"*\" * 50)\n",
    "print(f\"There are {len(vdjdb.obs.value_counts('antigen.species'))} antigen species\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bdc567f-d2dd-4c48-8211-a0124be7d5e9",
   "metadata": {},
   "source": [
    "### Calculate TCR distances to Database.\n",
    "\n",
    "- metric=’identity’: For now, we only consider exact sequence matches. We will later introduce queries for similar sequences as well.\n",
    "- sequence=’aa’: TCR sequences are compared on an amino-acid level instead of nucleobases, since specificity depends on the protein structure and this information is provided by the VDJdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "58e4def9-10aa-4d4f-ba58-399d1c37a459",
   "metadata": {},
   "outputs": [],
   "source": [
    "metric = \"identity\"\n",
    "sequence = \"aa\"\n",
    "ir.pp.ir_dist(tcell_mdata, vdjdb, metric=metric, sequence=sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5307cf24-413b-46bc-8e65-aba6441c1336",
   "metadata": {},
   "source": [
    "### Query Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35cbc2dc-20be-4f81-bdf2-f7bd781ed794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "15796d16b59140a2b39db1918fa751f0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1457 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sequence='aa'\n",
    "metric='identity'\n",
    "receptor_arm='VDJ'\n",
    "ir.tl.ir_query(\n",
    "    tcell_mdata[\"airr\"],\n",
    "    vdjdb,\n",
    "    metric=metric,\n",
    "    sequence=sequence,\n",
    "    receptor_arms=receptor_arm,\n",
    "    dual_ir=\"primary_only\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd2117ad-9d72-41a1-a42f-d936c1800930",
   "metadata": {},
   "source": [
    "# Calculate TCR distances\n",
    "\n",
    "To detect cells with shared specficity, we calculate the pairwise sequence distance between their TCRs.\n",
    "- We are doing this to group cells within a dataset to increase the number of hits we get when we query a TCR database.\n",
    "\n",
    "There are 3 ways to calculate the distances:\n",
    "1. **Edit Distances:** This method calculates the minimum number of operations required to transform one sequence into another. These operations typically include insertions, deletions, and substitutions of single elements in the sequence. For example, the edit distance between \"cat\" and \"hat\" is 1 because only one substitution is needed (replace 'c' with 'h'). This method is useful in understanding how similar two sequences are and in identifying the potential evolutionary changes between them.\n",
    "\n",
    "2. **k-mer Matching:** A k-mer is a substring of length 'k' from a longer sequence. The k-mer matching technique involves comparing the occurrence and frequency of these short motifs between two sequences. This method is often used for sequence alignment, genome assembly, and identifying sequence motifs that may have biological significance. For example, a 3-mer analysis of the sequence \"ATGCTAGC\" would involve looking at substrates like 'ATG', 'TGC', 'GCT', etc.\n",
    "\n",
    "3. **Embeddings:** In bioinformatics, embeddings often refer to the process of representing sequences as vectors in a high-dimensional space, usually through deep learning models. Each sequence gets transformed into a numeric vector, capturing essential features and relationships within the data. This approach is beneficial for tasks like sequence classification, clustering, and prediction. It's akin to how words are embedded in natural language processing (NLP) tasks, capturing semantic relationships between words.\n",
    "\n",
    "Focus on using either **TCRdist** or **TCRmatch**:\n",
    "\n",
    "**TCRdist**\n",
    "- uses the sequences of all CDRs and compares them via transformation cost and gap panelties [Dash et al., 2017].\n",
    "- TCRdist requires complete information of CDR3 and V-gene of both chains.\n",
    " \n",
    "**TCRmatch**\n",
    "- uses all k-mers to compare the overlap in motifs between two TCRs based on their CDR3β sequences\n",
    "    - Becuase it only requires CDR3β, we can use a wider array of data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cd22dbf-c737-43b8-a1bf-c4cb8686afa2",
   "metadata": {},
   "source": [
    "## TCRmatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "32adceca-c08b-492c-8909-422cb1bd051b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter cells with missing beta chains\n",
    "filter = [\"no IR\", \"multichain\", \"ambiguous\", \"orphan VJ\"]\n",
    "beta_tcr = mdata[mdata[\"airr\"].obs[\"chain_pairing\"].isin(filter)]\n",
    "\n",
    "# Keep only T cells\n",
    "not_tcells = [\"B intermediate\", \"Plasmablast\"]\n",
    "beta_tcr = beta_tcr[~beta_tcr[\"gex\"].obs[\"consensus_celltype\"].isin(not_tcells)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4b7b24a4-48c6-44f8-b67e-5f39013931f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample_Tag', 'receptor_type', 'receptor_subtype', 'chain_pairing',\n",
       "       'clone_id', 'clone_id_size', 'cc_aa_alignment', 'cc_aa_alignment_size',\n",
       "       'clone_id:full', 'clone_id_size:full', 'cc_aa_alignment:full',\n",
       "       'cc_aa_alignment_size:full', 'ID', 'clonal_expansion', 'is_full_chain'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beta_tcr[\"airr\"].obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "33b3cea4-e8cd-47e8-8a0b-6ca9cace13e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/anndata/_core/aligned_mapping.py:64: ExperimentalFeatureWarning: Support for Awkward Arrays is currently experimental. Behavior may change in the future. Please report any issues you may encounter!\n",
      "  warnings.warn(\n",
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:183: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>adt:Cell_Type_Experimental</th>\n",
       "      <th>adt:Sample_Tag</th>\n",
       "      <th>adt:Sample_Name</th>\n",
       "      <th>adt:SampleTag01_hs_Read_Count</th>\n",
       "      <th>adt:SampleTag02_hs_Read_Count</th>\n",
       "      <th>adt:SampleTag03_hs_Read_Count</th>\n",
       "      <th>adt:SampleTag04_hs_Read_Count</th>\n",
       "      <th>adt:SampleTag05_hs_Read_Count</th>\n",
       "      <th>adt:SampleTag06_hs_Read_Count</th>\n",
       "      <th>adt:SampleTag07_hs_Read_Count</th>\n",
       "      <th>...</th>\n",
       "      <th>airr:clone_id:full</th>\n",
       "      <th>airr:clone_id_size:full</th>\n",
       "      <th>airr:cc_aa_alignment:full</th>\n",
       "      <th>airr:cc_aa_alignment_size:full</th>\n",
       "      <th>airr:ID</th>\n",
       "      <th>airr:clonal_expansion</th>\n",
       "      <th>airr:is_full_chain</th>\n",
       "      <th>tcell_leiden</th>\n",
       "      <th>VDJ_1_v_call</th>\n",
       "      <th>VJ_1_v_call</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1218</th>\n",
       "      <td>T_CD4_memory</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV5*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6189</th>\n",
       "      <td>T_CD4_memory</td>\n",
       "      <td>SampleTag12_hs</td>\n",
       "      <td>SampleTag12_hs</td>\n",
       "      <td>17</td>\n",
       "      <td>64</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>44</td>\n",
       "      <td>205</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>13</td>\n",
       "      <td>TRBV2*01</td>\n",
       "      <td>IGKV4-1*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6572</th>\n",
       "      <td>T_CD8_memory</td>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>36</td>\n",
       "      <td>14</td>\n",
       "      <td>5262</td>\n",
       "      <td>103</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>TRBV12-3*01</td>\n",
       "      <td>TRGV2*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9608</th>\n",
       "      <td>T_CD4_memory</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>118</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21537</th>\n",
       "      <td>T_CD8_naive</td>\n",
       "      <td>SampleTag12_hs</td>\n",
       "      <td>SampleTag12_hs</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "      <td>62</td>\n",
       "      <td>31</td>\n",
       "      <td>84</td>\n",
       "      <td>84</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV29/DV5*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14033755</th>\n",
       "      <td>T_CD8_memory</td>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>21</td>\n",
       "      <td>133</td>\n",
       "      <td>12700</td>\n",
       "      <td>105</td>\n",
       "      <td>58</td>\n",
       "      <td>24</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>TRBV7-6*01</td>\n",
       "      <td>TRAV8-2*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14039104</th>\n",
       "      <td>T_CD4_memory</td>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>8381</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAV16*01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14041429</th>\n",
       "      <td>T_CD8_memory</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>30</td>\n",
       "      <td>51</td>\n",
       "      <td>19</td>\n",
       "      <td>65</td>\n",
       "      <td>8</td>\n",
       "      <td>47</td>\n",
       "      <td>69</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14042115</th>\n",
       "      <td>T_CD4_naive</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14042145</th>\n",
       "      <td>T_CD4_memory</td>\n",
       "      <td>SampleTag08_hs</td>\n",
       "      <td>SampleTag08_hs</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>954 rows × 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         adt:Cell_Type_Experimental  adt:Sample_Tag adt:Sample_Name  \\\n",
       "1218                   T_CD4_memory  SampleTag09_hs  SampleTag09_hs   \n",
       "6189                   T_CD4_memory  SampleTag12_hs  SampleTag12_hs   \n",
       "6572                   T_CD8_memory  SampleTag03_hs  SampleTag03_hs   \n",
       "9608                   T_CD4_memory  SampleTag09_hs  SampleTag09_hs   \n",
       "21537                   T_CD8_naive  SampleTag12_hs  SampleTag12_hs   \n",
       "...                             ...             ...             ...   \n",
       "14033755               T_CD8_memory  SampleTag03_hs  SampleTag03_hs   \n",
       "14039104               T_CD4_memory  SampleTag03_hs  SampleTag03_hs   \n",
       "14041429               T_CD8_memory  SampleTag09_hs  SampleTag09_hs   \n",
       "14042115                T_CD4_naive  SampleTag09_hs  SampleTag09_hs   \n",
       "14042145               T_CD4_memory  SampleTag08_hs  SampleTag08_hs   \n",
       "\n",
       "          adt:SampleTag01_hs_Read_Count  adt:SampleTag02_hs_Read_Count  \\\n",
       "1218                                  5                             15   \n",
       "6189                                 17                             64   \n",
       "6572                                 36                             14   \n",
       "9608                                  8                             16   \n",
       "21537                                13                             23   \n",
       "...                                 ...                            ...   \n",
       "14033755                             21                            133   \n",
       "14039104                             12                              3   \n",
       "14041429                             30                             51   \n",
       "14042115                              7                              6   \n",
       "14042145                              9                             28   \n",
       "\n",
       "          adt:SampleTag03_hs_Read_Count  adt:SampleTag04_hs_Read_Count  \\\n",
       "1218                                  7                             18   \n",
       "6189                                  8                              6   \n",
       "6572                               5262                            103   \n",
       "9608                                118                             42   \n",
       "21537                                 8                             62   \n",
       "...                                 ...                            ...   \n",
       "14033755                          12700                            105   \n",
       "14039104                           8381                             16   \n",
       "14041429                             19                             65   \n",
       "14042115                              0                             17   \n",
       "14042145                              0                             23   \n",
       "\n",
       "          adt:SampleTag05_hs_Read_Count  adt:SampleTag06_hs_Read_Count  \\\n",
       "1218                                  2                              2   \n",
       "6189                                 23                             44   \n",
       "6572                                  1                             29   \n",
       "9608                                  1                             41   \n",
       "21537                                31                             84   \n",
       "...                                 ...                            ...   \n",
       "14033755                             58                             24   \n",
       "14039104                             16                             10   \n",
       "14041429                              8                             47   \n",
       "14042115                              0                              9   \n",
       "14042145                              3                              9   \n",
       "\n",
       "          adt:SampleTag07_hs_Read_Count  ...  airr:clone_id:full  \\\n",
       "1218                                  6  ...                 NaN   \n",
       "6189                                205  ...                 NaN   \n",
       "6572                                 48  ...                 NaN   \n",
       "9608                                 35  ...                 NaN   \n",
       "21537                                84  ...                 NaN   \n",
       "...                                 ...  ...                 ...   \n",
       "14033755                             17  ...                 NaN   \n",
       "14039104                              1  ...                 NaN   \n",
       "14041429                             69  ...                 NaN   \n",
       "14042115                             14  ...                 NaN   \n",
       "14042145                             19  ...                 NaN   \n",
       "\n",
       "          airr:clone_id_size:full  airr:cc_aa_alignment:full  \\\n",
       "1218                          NaN                        NaN   \n",
       "6189                          NaN                        NaN   \n",
       "6572                          NaN                        NaN   \n",
       "9608                          NaN                        NaN   \n",
       "21537                         NaN                        NaN   \n",
       "...                           ...                        ...   \n",
       "14033755                      NaN                        NaN   \n",
       "14039104                      NaN                        NaN   \n",
       "14041429                      NaN                        NaN   \n",
       "14042115                      NaN                        NaN   \n",
       "14042145                      NaN                        NaN   \n",
       "\n",
       "          airr:cc_aa_alignment_size:full  airr:ID  airr:clonal_expansion  \\\n",
       "1218                                 NaN      NaN                    NaN   \n",
       "6189                                 NaN      NaN                    NaN   \n",
       "6572                                 NaN      NaN                    NaN   \n",
       "9608                                 NaN      NaN                    NaN   \n",
       "21537                                NaN      NaN                    NaN   \n",
       "...                                  ...      ...                    ...   \n",
       "14033755                             NaN      NaN                    NaN   \n",
       "14039104                             NaN      NaN                    NaN   \n",
       "14041429                             NaN      NaN                    NaN   \n",
       "14042115                             NaN      NaN                    NaN   \n",
       "14042145                             NaN      NaN                    NaN   \n",
       "\n",
       "          airr:is_full_chain tcell_leiden VDJ_1_v_call    VJ_1_v_call  \n",
       "1218                   False            4          NaN       TRAV5*01  \n",
       "6189                   False           13     TRBV2*01     IGKV4-1*01  \n",
       "6572                   False            2  TRBV12-3*01       TRGV2*01  \n",
       "9608                   False           22          NaN            NaN  \n",
       "21537                  False           10          NaN  TRAV29/DV5*01  \n",
       "...                      ...          ...          ...            ...  \n",
       "14033755               False            2   TRBV7-6*01     TRAV8-2*01  \n",
       "14039104               False           14          NaN      TRAV16*01  \n",
       "14041429               False            9          NaN            NaN  \n",
       "14042115               False            5          NaN            NaN  \n",
       "14042145               False            1          NaN            NaN  \n",
       "\n",
       "[954 rows x 168 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with ir.get.airr_context(beta_tcr, airr_variable=\"v_call\", chain=(\"VDJ_1\", \"VJ_1\")) as m:\n",
    "    display(m.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "659bb6f5-7fe5-4408-aee9-7abc054abd3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:760: ImplicitModificationWarning: Setting element `.varm['adt']` of view, initializing view as actual.\n",
      "  attrm[mod] = mapping > 0\n",
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:183: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:760: ImplicitModificationWarning: Setting element `.varm['gex']` of view, initializing view as actual.\n",
      "  attrm[mod] = mapping > 0\n",
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:183: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:760: ImplicitModificationWarning: Setting element `.varm['airr']` of view, initializing view as actual.\n",
      "  attrm[mod] = mapping > 0\n",
      "/Users/djlisko/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/anndata/_core/anndata.py:183: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"['airr:antigen.species'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 10\u001b[0m\n\u001b[1;32m      7\u001b[0m m \u001b[38;5;241m=\u001b[39m m[\u001b[38;5;241m~\u001b[39mm\u001b[38;5;241m.\u001b[39mobs[tcr_region]\u001b[38;5;241m.\u001b[39misna()]\n\u001b[1;32m      8\u001b[0m m\u001b[38;5;241m.\u001b[39mupdate()\n\u001b[0;32m---> 10\u001b[0m df_tcrdist \u001b[38;5;241m=\u001b[39m \u001b[43mm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobs\u001b[49m\u001b[43m[\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43m[\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mVDJ_1_junction_aa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mVDJ_1_v_call\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     14\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mVDJ_1_j_call\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mairr:chain_pairing\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgex:ID\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mairr:antigen.species\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgex:consensus_celltype\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m    \u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mcopy()\n",
      "File \u001b[0;32m~/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/pandas/core/frame.py:3767\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3765\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3766\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 3767\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3769\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3770\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/pandas/core/indexes/base.py:5877\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   5874\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   5875\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 5877\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5879\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   5880\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   5881\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/gitrepos/rhapsody-analysis-notebook/.venv/lib/python3.11/site-packages/pandas/core/indexes/base.py:5941\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   5938\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   5940\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[0;32m-> 5941\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['airr:antigen.species'] not in index\""
     ]
    }
   ],
   "source": [
    "tcr_region = \"VDJ_1_v_call\"\n",
    "cols = [\"VDJ_1_junction_aa\", \"VDJ_1_v_call\",  \"VDJ_1_j_call\"]\n",
    "\n",
    "# for col in [\"VJ_1_v_call\", \"VDJ_1_v_call\"]:\n",
    "\n",
    "with ir.get.airr_context(beta_tcr, [\"junction_aa\", \"v_call\", \"j_call\"]) as m:\n",
    "    m = m[~m.obs[tcr_region].isna()]\n",
    "    m.update()\n",
    "   \n",
    "    df_tcrdist = m.obs[\n",
    "        [\n",
    "            \"VDJ_1_junction_aa\",\n",
    "            \"VDJ_1_v_call\",\n",
    "            \"VDJ_1_j_call\",\n",
    "            \"airr:chain_pairing\",\n",
    "            \"gex:ID\",\n",
    "            \"airr:antigen.species\",\n",
    "            \"gex:consensus_celltype\",\n",
    "        ]\n",
    "    ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1e48b1ef-f069-4be2-ae52-d0c4716cb637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Tag</th>\n",
       "      <th>receptor_type</th>\n",
       "      <th>receptor_subtype</th>\n",
       "      <th>chain_pairing</th>\n",
       "      <th>clone_id</th>\n",
       "      <th>clone_id_size</th>\n",
       "      <th>cc_aa_alignment</th>\n",
       "      <th>cc_aa_alignment_size</th>\n",
       "      <th>clone_id:full</th>\n",
       "      <th>clone_id_size:full</th>\n",
       "      <th>cc_aa_alignment:full</th>\n",
       "      <th>cc_aa_alignment_size:full</th>\n",
       "      <th>ID</th>\n",
       "      <th>clonal_expansion</th>\n",
       "      <th>is_full_chain</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1218</th>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>TCR</td>\n",
       "      <td>TRA+TRB</td>\n",
       "      <td>orphan VJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6189</th>\n",
       "      <td>SampleTag12_hs</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6572</th>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>TCR</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>8</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9608</th>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21537</th>\n",
       "      <td>SampleTag12_hs</td>\n",
       "      <td>TCR</td>\n",
       "      <td>TRA+TRB</td>\n",
       "      <td>orphan VJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14033755</th>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>TCR</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>ambiguous</td>\n",
       "      <td>1583</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14039104</th>\n",
       "      <td>SampleTag03_hs</td>\n",
       "      <td>TCR</td>\n",
       "      <td>TRA+TRB</td>\n",
       "      <td>orphan VJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14041429</th>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14042115</th>\n",
       "      <td>SampleTag09_hs</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14042145</th>\n",
       "      <td>SampleTag08_hs</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>no IR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>954 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Sample_Tag receptor_type receptor_subtype chain_pairing  \\\n",
       "cell_id                                                                 \n",
       "1218      SampleTag09_hs           TCR          TRA+TRB     orphan VJ   \n",
       "6189      SampleTag12_hs     ambiguous        ambiguous     ambiguous   \n",
       "6572      SampleTag03_hs           TCR        ambiguous     ambiguous   \n",
       "9608      SampleTag09_hs         no IR            no IR         no IR   \n",
       "21537     SampleTag12_hs           TCR          TRA+TRB     orphan VJ   \n",
       "...                  ...           ...              ...           ...   \n",
       "14033755  SampleTag03_hs           TCR        ambiguous     ambiguous   \n",
       "14039104  SampleTag03_hs           TCR          TRA+TRB     orphan VJ   \n",
       "14041429  SampleTag09_hs         no IR            no IR         no IR   \n",
       "14042115  SampleTag09_hs         no IR            no IR         no IR   \n",
       "14042145  SampleTag08_hs         no IR            no IR         no IR   \n",
       "\n",
       "         clone_id  clone_id_size cc_aa_alignment  cc_aa_alignment_size  \\\n",
       "cell_id                                                                  \n",
       "1218          NaN            NaN             NaN                   NaN   \n",
       "6189            7            1.0               7                   1.0   \n",
       "6572            8            3.0               8                   3.0   \n",
       "9608          NaN            NaN             NaN                   NaN   \n",
       "21537         NaN            NaN             NaN                   NaN   \n",
       "...           ...            ...             ...                   ...   \n",
       "14033755     1583            1.0             516                   2.0   \n",
       "14039104      NaN            NaN             NaN                   NaN   \n",
       "14041429      NaN            NaN             NaN                   NaN   \n",
       "14042115      NaN            NaN             NaN                   NaN   \n",
       "14042145      NaN            NaN             NaN                   NaN   \n",
       "\n",
       "         clone_id:full  clone_id_size:full cc_aa_alignment:full  \\\n",
       "cell_id                                                           \n",
       "1218               NaN                 NaN                  NaN   \n",
       "6189               NaN                 NaN                  NaN   \n",
       "6572               NaN                 NaN                  NaN   \n",
       "9608               NaN                 NaN                  NaN   \n",
       "21537              NaN                 NaN                  NaN   \n",
       "...                ...                 ...                  ...   \n",
       "14033755           NaN                 NaN                  NaN   \n",
       "14039104           NaN                 NaN                  NaN   \n",
       "14041429           NaN                 NaN                  NaN   \n",
       "14042115           NaN                 NaN                  NaN   \n",
       "14042145           NaN                 NaN                  NaN   \n",
       "\n",
       "          cc_aa_alignment_size:full   ID clonal_expansion  is_full_chain  \n",
       "cell_id                                                                   \n",
       "1218                            NaN  NaN              NaN          False  \n",
       "6189                            NaN  NaN              NaN          False  \n",
       "6572                            NaN  NaN              NaN          False  \n",
       "9608                            NaN  NaN              NaN          False  \n",
       "21537                           NaN  NaN              NaN          False  \n",
       "...                             ...  ...              ...            ...  \n",
       "14033755                        NaN  NaN              NaN          False  \n",
       "14039104                        NaN  NaN              NaN          False  \n",
       "14041429                        NaN  NaN              NaN          False  \n",
       "14042115                        NaN  NaN              NaN          False  \n",
       "14042145                        NaN  NaN              NaN          False  \n",
       "\n",
       "[954 rows x 15 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beta_tcr[\"airr\"].obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "88a01de6-c885-4040-806f-37c39a5749cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 2822 × 4021\n",
       "  obs:\t&#x27;tcell_leiden&#x27;\n",
       "  var:\t&#x27;Raw_Reads&#x27;, &#x27;Raw_Molecules&#x27;, &#x27;Raw_Seq_Depth&#x27;, &#x27;RSEC_Adjusted_Molecules&#x27;, &#x27;RSEC_Adjusted_Reads_non-singleton&#x27;, &#x27;RSEC_Adjusted_Molecules_non-singleton&#x27;, &#x27;Gene_Expression&#x27;, &#x27;n_cells_by_counts&#x27;, &#x27;mean_counts&#x27;, &#x27;log1p_mean_counts&#x27;, &#x27;pct_dropout_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;\n",
       "  3 modalities\n",
       "    adt:\t2822 x 21\n",
       "      obs:\t&#x27;Cell_Type_Experimental&#x27;, &#x27;Sample_Tag&#x27;, &#x27;Sample_Name&#x27;, &#x27;SampleTag01_hs_Read_Count&#x27;, &#x27;SampleTag02_hs_Read_Count&#x27;, &#x27;SampleTag03_hs_Read_Count&#x27;, &#x27;SampleTag04_hs_Read_Count&#x27;, &#x27;SampleTag05_hs_Read_Count&#x27;, &#x27;SampleTag06_hs_Read_Count&#x27;, &#x27;SampleTag07_hs_Read_Count&#x27;, &#x27;SampleTag08_hs_Read_Count&#x27;, &#x27;SampleTag09_hs_Read_Count&#x27;, &#x27;SampleTag10_hs_Read_Count&#x27;, &#x27;SampleTag11_hs_Read_Count&#x27;, &#x27;SampleTag12_hs_Read_Count&#x27;, &#x27;Total_VDJ_Read_Count&#x27;, &#x27;Total_VDJ_Molecule_Count&#x27;, &#x27;BCR_Heavy_V_gene_Dominant&#x27;, &#x27;BCR_Heavy_D_gene_Dominant&#x27;, &#x27;BCR_Heavy_J_gene_Dominant&#x27;, &#x27;BCR_Heavy_C_gene_Dominant&#x27;, &#x27;BCR_Heavy_CDR3_Nucleotide_Dominant&#x27;, &#x27;BCR_Heavy_CDR3_Translation_Dominant&#x27;, &#x27;BCR_Heavy_Read_Count&#x27;, &#x27;BCR_Heavy_Molecule_Count&#x27;, &#x27;BCR_Light_V_gene_Dominant&#x27;, &#x27;BCR_Light_J_gene_Dominant&#x27;, &#x27;BCR_Light_C_gene_Dominant&#x27;, &#x27;BCR_Light_CDR3_Nucleotide_Dominant&#x27;, &#x27;BCR_Light_CDR3_Translation_Dominant&#x27;, &#x27;BCR_Light_Read_Count&#x27;, &#x27;BCR_Light_Molecule_Count&#x27;, &#x27;TCR_Alpha_Gamma_V_gene_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_J_gene_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_C_gene_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_CDR3_Nucleotide_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_CDR3_Translation_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_Read_Count&#x27;, &#x27;TCR_Alpha_Gamma_Molecule_Count&#x27;, &#x27;TCR_Beta_Delta_V_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_D_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_J_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_C_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_CDR3_Nucleotide_Dominant&#x27;, &#x27;TCR_Beta_Delta_CDR3_Translation_Dominant&#x27;, &#x27;TCR_Beta_Delta_Read_Count&#x27;, &#x27;TCR_Beta_Delta_Molecule_Count&#x27;, &#x27;BCR_Paired_Chains&#x27;, &#x27;TCR_Paired_Chains&#x27;, &#x27;ID&#x27;, &#x27;samle_number&#x27;, &#x27;day&#x27;, &#x27;date&#x27;, &#x27;time_points&#x27;, &#x27;sample_type&#x27;, &#x27;aghvd&#x27;, &#x27;n_genes_by_counts&#x27;, &#x27;log1p_n_genes_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;n_counts&#x27;\n",
       "      var:\t&#x27;Raw_Reads&#x27;, &#x27;Raw_Molecules&#x27;, &#x27;Raw_Seq_Depth&#x27;, &#x27;RSEC_Adjusted_Molecules&#x27;, &#x27;RSEC_Adjusted_Reads_non-singleton&#x27;, &#x27;RSEC_Adjusted_Molecules_non-singleton&#x27;, &#x27;Gene_Expression&#x27;, &#x27;n_cells_by_counts&#x27;, &#x27;mean_counts&#x27;, &#x27;log1p_mean_counts&#x27;, &#x27;pct_dropout_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;adt_names&#x27;\n",
       "      uns:\t&#x27;Pipeline_Inputs&#x27;, &#x27;Pipeline_Metrics&#x27;\n",
       "      obsm:\t&#x27;X_tsne&#x27;\n",
       "      layers:\t&#x27;counts&#x27;\n",
       "    gex:\t2822 x 4000\n",
       "      obs:\t&#x27;Cell_Type_Experimental&#x27;, &#x27;Sample_Tag&#x27;, &#x27;Sample_Name&#x27;, &#x27;SampleTag01_hs_Read_Count&#x27;, &#x27;SampleTag02_hs_Read_Count&#x27;, &#x27;SampleTag03_hs_Read_Count&#x27;, &#x27;SampleTag04_hs_Read_Count&#x27;, &#x27;SampleTag05_hs_Read_Count&#x27;, &#x27;SampleTag06_hs_Read_Count&#x27;, &#x27;SampleTag07_hs_Read_Count&#x27;, &#x27;SampleTag08_hs_Read_Count&#x27;, &#x27;SampleTag09_hs_Read_Count&#x27;, &#x27;SampleTag10_hs_Read_Count&#x27;, &#x27;SampleTag11_hs_Read_Count&#x27;, &#x27;SampleTag12_hs_Read_Count&#x27;, &#x27;Total_VDJ_Read_Count&#x27;, &#x27;Total_VDJ_Molecule_Count&#x27;, &#x27;BCR_Heavy_V_gene_Dominant&#x27;, &#x27;BCR_Heavy_D_gene_Dominant&#x27;, &#x27;BCR_Heavy_J_gene_Dominant&#x27;, &#x27;BCR_Heavy_C_gene_Dominant&#x27;, &#x27;BCR_Heavy_CDR3_Nucleotide_Dominant&#x27;, &#x27;BCR_Heavy_CDR3_Translation_Dominant&#x27;, &#x27;BCR_Heavy_Read_Count&#x27;, &#x27;BCR_Heavy_Molecule_Count&#x27;, &#x27;BCR_Light_V_gene_Dominant&#x27;, &#x27;BCR_Light_J_gene_Dominant&#x27;, &#x27;BCR_Light_C_gene_Dominant&#x27;, &#x27;BCR_Light_CDR3_Nucleotide_Dominant&#x27;, &#x27;BCR_Light_CDR3_Translation_Dominant&#x27;, &#x27;BCR_Light_Read_Count&#x27;, &#x27;BCR_Light_Molecule_Count&#x27;, &#x27;TCR_Alpha_Gamma_V_gene_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_J_gene_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_C_gene_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_CDR3_Nucleotide_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_CDR3_Translation_Dominant&#x27;, &#x27;TCR_Alpha_Gamma_Read_Count&#x27;, &#x27;TCR_Alpha_Gamma_Molecule_Count&#x27;, &#x27;TCR_Beta_Delta_V_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_D_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_J_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_C_gene_Dominant&#x27;, &#x27;TCR_Beta_Delta_CDR3_Nucleotide_Dominant&#x27;, &#x27;TCR_Beta_Delta_CDR3_Translation_Dominant&#x27;, &#x27;TCR_Beta_Delta_Read_Count&#x27;, &#x27;TCR_Beta_Delta_Molecule_Count&#x27;, &#x27;BCR_Paired_Chains&#x27;, &#x27;TCR_Paired_Chains&#x27;, &#x27;ID&#x27;, &#x27;samle_number&#x27;, &#x27;day&#x27;, &#x27;date&#x27;, &#x27;time_points&#x27;, &#x27;sample_type&#x27;, &#x27;aghvd&#x27;, &#x27;n_genes_by_counts&#x27;, &#x27;log1p_n_genes_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;pct_counts_in_top_20_genes&#x27;, &#x27;total_counts_mt&#x27;, &#x27;log1p_total_counts_mt&#x27;, &#x27;pct_counts_mt&#x27;, &#x27;total_counts_ribo&#x27;, &#x27;log1p_total_counts_ribo&#x27;, &#x27;pct_counts_ribo&#x27;, &#x27;total_counts_hb&#x27;, &#x27;log1p_total_counts_hb&#x27;, &#x27;pct_counts_hb&#x27;, &#x27;doublet_class&#x27;, &#x27;doublet_score&#x27;, &#x27;exp_cell_l1&#x27;, &#x27;az_l1_score&#x27;, &#x27;az_l1_celltype&#x27;, &#x27;az_l2_score&#x27;, &#x27;az_l2_celltype&#x27;, &#x27;ct_low_predicted_labels&#x27;, &#x27;ct_low_over_clustering&#x27;, &#x27;ct_low_majority_voting&#x27;, &#x27;ct_low_conf_score&#x27;, &#x27;ct_high_predicted_labels&#x27;, &#x27;ct_high_over_clustering&#x27;, &#x27;ct_high_majority_voting&#x27;, &#x27;ct_high_conf_score&#x27;, &#x27;abt_label&#x27;, &#x27;is_gdT&#x27;, &#x27;consensus_celltype&#x27;, &#x27;day_label&#x27;\n",
       "      var:\t&#x27;Raw_Reads&#x27;, &#x27;Raw_Molecules&#x27;, &#x27;Raw_Seq_Depth&#x27;, &#x27;RSEC_Adjusted_Molecules&#x27;, &#x27;RSEC_Adjusted_Reads_non-singleton&#x27;, &#x27;RSEC_Adjusted_Molecules_non-singleton&#x27;, &#x27;Gene_Expression&#x27;, &#x27;mt&#x27;, &#x27;ribo&#x27;, &#x27;hb&#x27;, &#x27;n_cells_by_counts&#x27;, &#x27;mean_counts&#x27;, &#x27;log1p_mean_counts&#x27;, &#x27;pct_dropout_by_counts&#x27;, &#x27;total_counts&#x27;, &#x27;log1p_total_counts&#x27;, &#x27;binomial_dev&#x27;, &#x27;highly_deviant&#x27;, &#x27;binomial_deviance&#x27;, &#x27;highly_variable&#x27;, &#x27;means&#x27;, &#x27;dispersions&#x27;, &#x27;dispersions_norm&#x27;, &#x27;mean&#x27;, &#x27;std&#x27;\n",
       "      uns:\t&#x27;Pipeline_Inputs&#x27;, &#x27;Pipeline_Metrics&#x27;, &#x27;Sample_Tag_colors&#x27;, &#x27;az_l1_celltype_colors&#x27;, &#x27;az_l2_celltype_colors&#x27;, &#x27;exp_cell_l1_colors&#x27;, &#x27;gex:az_l2_celltype_colors&#x27;, &#x27;gex:is_gdT_colors&#x27;, &#x27;hvg&#x27;, &#x27;neighbors&#x27;, &#x27;pca&#x27;, &#x27;umap&#x27;\n",
       "      obsm:\t&#x27;X_pca&#x27;, &#x27;X_tsne&#x27;, &#x27;X_umap&#x27;\n",
       "      varm:\t&#x27;PCs&#x27;\n",
       "      layers:\t&#x27;log1p_norm&#x27;, &#x27;raw&#x27;\n",
       "      obsp:\t&#x27;connectivities&#x27;, &#x27;distances&#x27;\n",
       "    airr:\t2822 x 0\n",
       "      obs:\t&#x27;Sample_Tag&#x27;, &#x27;receptor_type&#x27;, &#x27;receptor_subtype&#x27;, &#x27;chain_pairing&#x27;, &#x27;clone_id&#x27;, &#x27;clone_id_size&#x27;, &#x27;cc_aa_alignment&#x27;, &#x27;cc_aa_alignment_size&#x27;, &#x27;clone_id:full&#x27;, &#x27;clone_id_size:full&#x27;, &#x27;cc_aa_alignment:full&#x27;, &#x27;cc_aa_alignment_size:full&#x27;, &#x27;ID&#x27;, &#x27;clonal_expansion&#x27;, &#x27;is_full_chain&#x27;\n",
       "      uns:\t&#x27;cc_aa_alignment&#x27;, &#x27;chain_indices&#x27;, &#x27;clone_id&#x27;, &#x27;clonotype_network&#x27;, &#x27;ir_dist_aa_alignment&#x27;, &#x27;ir_dist_nt_identity&#x27;, &#x27;scirpy_version&#x27;, &#x27;ir_dist_VDJDB_aa_identity&#x27;, &#x27;ir_query_VDJDB_aa_identity&#x27;\n",
       "      obsm:\t&#x27;X_clonotype_network&#x27;, &#x27;airr&#x27;, &#x27;chain_indices&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 2822 × 4021\n",
       "  obs:\t'tcell_leiden'\n",
       "  var:\t'Raw_Reads', 'Raw_Molecules', 'Raw_Seq_Depth', 'RSEC_Adjusted_Molecules', 'RSEC_Adjusted_Reads_non-singleton', 'RSEC_Adjusted_Molecules_non-singleton', 'Gene_Expression', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "  3 modalities\n",
       "    adt:\t2822 x 21\n",
       "      obs:\t'Cell_Type_Experimental', 'Sample_Tag', 'Sample_Name', 'SampleTag01_hs_Read_Count', 'SampleTag02_hs_Read_Count', 'SampleTag03_hs_Read_Count', 'SampleTag04_hs_Read_Count', 'SampleTag05_hs_Read_Count', 'SampleTag06_hs_Read_Count', 'SampleTag07_hs_Read_Count', 'SampleTag08_hs_Read_Count', 'SampleTag09_hs_Read_Count', 'SampleTag10_hs_Read_Count', 'SampleTag11_hs_Read_Count', 'SampleTag12_hs_Read_Count', 'Total_VDJ_Read_Count', 'Total_VDJ_Molecule_Count', 'BCR_Heavy_V_gene_Dominant', 'BCR_Heavy_D_gene_Dominant', 'BCR_Heavy_J_gene_Dominant', 'BCR_Heavy_C_gene_Dominant', 'BCR_Heavy_CDR3_Nucleotide_Dominant', 'BCR_Heavy_CDR3_Translation_Dominant', 'BCR_Heavy_Read_Count', 'BCR_Heavy_Molecule_Count', 'BCR_Light_V_gene_Dominant', 'BCR_Light_J_gene_Dominant', 'BCR_Light_C_gene_Dominant', 'BCR_Light_CDR3_Nucleotide_Dominant', 'BCR_Light_CDR3_Translation_Dominant', 'BCR_Light_Read_Count', 'BCR_Light_Molecule_Count', 'TCR_Alpha_Gamma_V_gene_Dominant', 'TCR_Alpha_Gamma_J_gene_Dominant', 'TCR_Alpha_Gamma_C_gene_Dominant', 'TCR_Alpha_Gamma_CDR3_Nucleotide_Dominant', 'TCR_Alpha_Gamma_CDR3_Translation_Dominant', 'TCR_Alpha_Gamma_Read_Count', 'TCR_Alpha_Gamma_Molecule_Count', 'TCR_Beta_Delta_V_gene_Dominant', 'TCR_Beta_Delta_D_gene_Dominant', 'TCR_Beta_Delta_J_gene_Dominant', 'TCR_Beta_Delta_C_gene_Dominant', 'TCR_Beta_Delta_CDR3_Nucleotide_Dominant', 'TCR_Beta_Delta_CDR3_Translation_Dominant', 'TCR_Beta_Delta_Read_Count', 'TCR_Beta_Delta_Molecule_Count', 'BCR_Paired_Chains', 'TCR_Paired_Chains', 'ID', 'samle_number', 'day', 'date', 'time_points', 'sample_type', 'aghvd', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'n_counts'\n",
       "      var:\t'Raw_Reads', 'Raw_Molecules', 'Raw_Seq_Depth', 'RSEC_Adjusted_Molecules', 'RSEC_Adjusted_Reads_non-singleton', 'RSEC_Adjusted_Molecules_non-singleton', 'Gene_Expression', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'adt_names'\n",
       "      uns:\t'Pipeline_Inputs', 'Pipeline_Metrics'\n",
       "      obsm:\t'X_tsne'\n",
       "      layers:\t'counts'\n",
       "    gex:\t2822 x 4000\n",
       "      obs:\t'Cell_Type_Experimental', 'Sample_Tag', 'Sample_Name', 'SampleTag01_hs_Read_Count', 'SampleTag02_hs_Read_Count', 'SampleTag03_hs_Read_Count', 'SampleTag04_hs_Read_Count', 'SampleTag05_hs_Read_Count', 'SampleTag06_hs_Read_Count', 'SampleTag07_hs_Read_Count', 'SampleTag08_hs_Read_Count', 'SampleTag09_hs_Read_Count', 'SampleTag10_hs_Read_Count', 'SampleTag11_hs_Read_Count', 'SampleTag12_hs_Read_Count', 'Total_VDJ_Read_Count', 'Total_VDJ_Molecule_Count', 'BCR_Heavy_V_gene_Dominant', 'BCR_Heavy_D_gene_Dominant', 'BCR_Heavy_J_gene_Dominant', 'BCR_Heavy_C_gene_Dominant', 'BCR_Heavy_CDR3_Nucleotide_Dominant', 'BCR_Heavy_CDR3_Translation_Dominant', 'BCR_Heavy_Read_Count', 'BCR_Heavy_Molecule_Count', 'BCR_Light_V_gene_Dominant', 'BCR_Light_J_gene_Dominant', 'BCR_Light_C_gene_Dominant', 'BCR_Light_CDR3_Nucleotide_Dominant', 'BCR_Light_CDR3_Translation_Dominant', 'BCR_Light_Read_Count', 'BCR_Light_Molecule_Count', 'TCR_Alpha_Gamma_V_gene_Dominant', 'TCR_Alpha_Gamma_J_gene_Dominant', 'TCR_Alpha_Gamma_C_gene_Dominant', 'TCR_Alpha_Gamma_CDR3_Nucleotide_Dominant', 'TCR_Alpha_Gamma_CDR3_Translation_Dominant', 'TCR_Alpha_Gamma_Read_Count', 'TCR_Alpha_Gamma_Molecule_Count', 'TCR_Beta_Delta_V_gene_Dominant', 'TCR_Beta_Delta_D_gene_Dominant', 'TCR_Beta_Delta_J_gene_Dominant', 'TCR_Beta_Delta_C_gene_Dominant', 'TCR_Beta_Delta_CDR3_Nucleotide_Dominant', 'TCR_Beta_Delta_CDR3_Translation_Dominant', 'TCR_Beta_Delta_Read_Count', 'TCR_Beta_Delta_Molecule_Count', 'BCR_Paired_Chains', 'TCR_Paired_Chains', 'ID', 'samle_number', 'day', 'date', 'time_points', 'sample_type', 'aghvd', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'doublet_class', 'doublet_score', 'exp_cell_l1', 'az_l1_score', 'az_l1_celltype', 'az_l2_score', 'az_l2_celltype', 'ct_low_predicted_labels', 'ct_low_over_clustering', 'ct_low_majority_voting', 'ct_low_conf_score', 'ct_high_predicted_labels', 'ct_high_over_clustering', 'ct_high_majority_voting', 'ct_high_conf_score', 'abt_label', 'is_gdT', 'consensus_celltype', 'day_label'\n",
       "      var:\t'Raw_Reads', 'Raw_Molecules', 'Raw_Seq_Depth', 'RSEC_Adjusted_Molecules', 'RSEC_Adjusted_Reads_non-singleton', 'RSEC_Adjusted_Molecules_non-singleton', 'Gene_Expression', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'binomial_dev', 'highly_deviant', 'binomial_deviance', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "      uns:\t'Pipeline_Inputs', 'Pipeline_Metrics', 'Sample_Tag_colors', 'az_l1_celltype_colors', 'az_l2_celltype_colors', 'exp_cell_l1_colors', 'gex:az_l2_celltype_colors', 'gex:is_gdT_colors', 'hvg', 'neighbors', 'pca', 'umap'\n",
       "      obsm:\t'X_pca', 'X_tsne', 'X_umap'\n",
       "      varm:\t'PCs'\n",
       "      layers:\t'log1p_norm', 'raw'\n",
       "      obsp:\t'connectivities', 'distances'\n",
       "    airr:\t2822 x 0\n",
       "      obs:\t'Sample_Tag', 'receptor_type', 'receptor_subtype', 'chain_pairing', 'clone_id', 'clone_id_size', 'cc_aa_alignment', 'cc_aa_alignment_size', 'clone_id:full', 'clone_id_size:full', 'cc_aa_alignment:full', 'cc_aa_alignment_size:full', 'ID', 'clonal_expansion', 'is_full_chain'\n",
       "      uns:\t'cc_aa_alignment', 'chain_indices', 'clone_id', 'clonotype_network', 'ir_dist_aa_alignment', 'ir_dist_nt_identity', 'scirpy_version', 'ir_dist_VDJDB_aa_identity', 'ir_query_VDJDB_aa_identity'\n",
       "      obsm:\t'X_clonotype_network', 'airr', 'chain_indices'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcell_mdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db3ef46c-35a4-49da-ab14-1a356f403da0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
